<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Reconstruct and analyse multi-omics networks with carmon • carmon</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Reconstruct and analyse multi-omics networks with carmon">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">carmon</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/carmon.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/DrQuestion/carmon/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Reconstruct and analyse multi-omics networks with carmon</h1>
                        <h4 data-toc-skip class="author">Alessio
Albanese</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/DrQuestion/carmon/blob/main/vignettes/carmon.Rmd" class="external-link"><code>vignettes/carmon.Rmd</code></a></small>
      <div class="d-none name"><code>carmon.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Copula-Aided Reconstruction of Multi-Omics Networks () is an R
package empowered by the use of copulas as a statistical tool for the
reconstruction of multi-omics networks. The use of copulas as a
statistical tool allows to transfer raw non-normalized omics data to the
normal realm. In this way, the package is able to harness the full
statistical power of the information contained in each omics layer of a
multi-omics data set. At the same time, the transfer of the multi-omics
layer to the normal realm allows the application of network
reconstruction models that assume normal distribution of the data.
Needless to say, this is an assumption rarely met by omics data, hence a
further advantage of the use of copulas. Here, we first cover the
formatting of the input data to provide to . Then, we show how to easily
build and analyse a multi-omics network with carmon. In the remaining
sections, we explore more advanced and customized uses of the package,
as well as what actually happens under the hood when the main function
<code><a href="../reference/carmon.html">carmon()</a></code> is used.</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>You can install from <a href="https://bioconductor.org/" class="external-link">Bioconductor</a> with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.github.io/BiocManager/" class="external-link">"BiocManager"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"carmon"</span><span class="op">)</span></span></code></pre></div>
<p>Alternatively, you can install the <em>development</em> version from
<a href="https://github.com/" class="external-link">GitHub</a>. First, make sure to install
devtools with:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://devtools.r-lib.org/" class="external-link">"devtools"</a></span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>You can then install the development version with:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"DrQuestion/carmon"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="input-multi-omics-data-set-up">Input multi-omics data set-up<a class="anchor" aria-label="anchor" href="#input-multi-omics-data-set-up"></a>
</h2>
<p>We first attach the package.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/DrQuestion/carmon" class="external-link">carmon</a></span><span class="op">)</span></span></code></pre></div>
<p>We can now explain the data format. Carmon prefers the input
multi-omics data to be arranged in a named R list. Each element of the
list should be the data set of one of the omics layers, and it should be
preferably named after the omics type it contains. In the package there
are several examples of multi-omics data sets of increasing sizes, which
you can see with <code><a href="../reference/multi_omics.html">help(multi_omics)</a></code>. Here we use
<code>multi_omics_small</code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">multi_omics_small</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">multi_omics_small</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 2</span></span>
<span><span class="co">#&gt;  $ rnaseq      : num [1:30, 1:14] 1023 815 1047 725 1217 ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:30] "BXD100" "BXD101" "BXD73b" "BXD44" ...</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:14] "Cirbp" "Hspa5" "P4ha1" "Spred1" ...</span></span>
<span><span class="co">#&gt;  $ metabolomics: num [1:30, 1:5] 66.1 50.2 56.3 58.9 61 ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:30] "BXD100" "BXD101" "BXD73b" "BXD44" ...</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:5] "Phe" "Trp" "Putrescine" "PC aa C36:3" ...</span></span></code></pre></div>
<p>You can see that this list has two elements, one containing RNA-seq
gene counts and named “rnaseq” (<code>multi_omics_small$rnaseq</code>)
and the other containing metabolites concentrations and named
“metabolomics” (<code>multi_omics_small$metabolomics</code>). Each layer
has 30 observations, the first one measuring the expression of 14 genes,
the second one the concentration of 5 metabolites. Please notice that
carmon expects non-normalized data, to harness at the best the
statistical behavior of omics data with copula.</p>
<p>To recreate such an object on R, first it is necessary to make sure
that the <strong>observations/samples/individuals</strong> from which
the data set has been measured <strong>are in the same order for all the
input omics data sets</strong> and <strong>are consistent in where they
are placed for all data sets,</strong> either <em>all</em> across the
rows or <em>all</em> across the column. For example, in our toy data
set, you can see how all the observations are in the same order and
distributed along the rows for both the transcriptomic and the
metabolomic data set. After your separate data sets are properly
formatted, you can assemble them in a single list as follows:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Let's build two separate data sets just for the sake of our example:</span></span>
<span><span class="va">layer_1</span> <span class="op">&lt;-</span> <span class="va">multi_omics_small</span><span class="op">$</span><span class="va">rnaseq</span></span>
<span><span class="va">layer_2</span> <span class="op">&lt;-</span> <span class="va">multi_omics_small</span><span class="op">$</span><span class="va">metabolomics</span></span>
<span></span>
<span><span class="co"># This shows that observations are on the rows for both data sets:</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">layer_1</span><span class="op">)</span> <span class="co"># [1] 30 14</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">layer_2</span><span class="op">)</span> <span class="co"># [1] 30  5</span></span>
<span></span>
<span><span class="co"># This shows that the observations are all in the same order in both data sets:</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">layer_1</span><span class="op">)</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">layer_2</span><span class="op">)</span><span class="op">)</span> <span class="co"># TRUE</span></span>
<span></span>
<span><span class="co"># We can now build our R list as follows:</span></span>
<span><span class="va">multi_omics_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"rnaseq"</span> <span class="op">=</span> <span class="va">layer_1</span>, <span class="st">"metabolomics"</span> <span class="op">=</span> <span class="va">layer_2</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="using-multiassayexperiment-objects-or-single-unified-data-sets-as-input">Using <code>MultiAssayExperiment</code> objects or single unified
data sets as input<a class="anchor" aria-label="anchor" href="#using-multiassayexperiment-objects-or-single-unified-data-sets-as-input"></a>
</h3>
<p>It is also possible to provide <code><a href="../reference/carmon.html">carmon()</a></code> with a <a href="https://doi.org/doi:10.18129/B9.bioc.MultiAssayExperiment" class="external-link">MultiAssayExperiment</a>
or a single unified multi-omics data set. In the latter, it is necessary
to specify the argument <code>p</code>, a vector with the number of
features of each layer, in the same order as the layers are assembled in
the data set. Moreover, with both these use cases, we strongly recommend
using the <code>omics</code> argument to specify which omics type is
measured by each layer. To see which omics types <code><a href="../reference/carmon.html">carmon()</a></code>
is tailored for, together with their accepted synonyms and assumed
statistical behavior, see <code><a href="../reference/which_omics.html">which_omics()</a></code>.</p>
</div>
</div>
<div class="section level2">
<h2 id="carmon-to-perform-copula-aided-reconstruction-analysis-and-plot-of-a-multi-omics-network">Carmon to perform copula-aided reconstruction, analysis and plot of
a <!-- 
--> multi-omics network<a class="anchor" aria-label="anchor" href="#carmon-to-perform-copula-aided-reconstruction-analysis-and-plot-of-a-multi-omics-network"></a>
</h2>
<p>The carmon package covers multiple functional steps of network
reconstruction and analysis from multi-omics data. To perform a full,
default run of carmon, it is necessary to run:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">c_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/carmon.html">carmon</a></span><span class="op">(</span><span class="va">multi_omics_small</span><span class="op">)</span></span>
<span><span class="co">#&gt; Checking sample-matching, formatting the data,</span></span>
<span><span class="co">#&gt;     and other formalities....</span></span>
<span><span class="co">#&gt; Done!</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ****************Beginning copulization****************</span></span>
<span><span class="co">#&gt; Copulizing layer 1 of 2 (rnaseq)</span></span>
<span><span class="co">#&gt; Copulizing layer 2 of 2 (metabolomics)</span></span>
<span><span class="co">#&gt; ****************Copulization complete*****************</span></span>
<span><span class="co">#&gt; ***********Beginning network reconstruction***********</span></span>
<span><span class="co">#&gt; Reconstructing network with coglasso, selecting the optimal one with xestars....</span></span>
<span><span class="co">#&gt; Optimal network selected!</span></span>
<span><span class="co">#&gt; ***********Network reconstruction complete************</span></span>
<span><span class="co">#&gt; **************Beginning network analysis**************</span></span>
<span><span class="co">#&gt; Computing centrality measures....</span></span>
<span><span class="co">#&gt; Centrality measures computed!</span></span>
<span><span class="co">#&gt; Compiling the report....</span></span>
<span><span class="co">#&gt; Report compiled!</span></span>
<span><span class="co">#&gt; **************Network analysis complete***************</span></span></code></pre></div>
<p><img src="carmon_files/figure-html/default%20run-1.png" width="576"><img src="carmon_files/figure-html/default%20run-2.png" width="576"></p>
<p>Under the hood, there are several processes happening. In order: the
check-ups and formatting of the input data; the copula-based transfer of
non-normal omics data to the normal realm; the reconstruction and
selection of a multi-omics network; a centrality consensus analysis to
identify key components of the network; and the plotting of the
multi-omics network (enriched by the information discovered during the
centrality analysis) and of the results of the centrality analysis. Most
of these steps can be customized within the main wrapper function, using
specific arguments. We will see how in the next sections, but first
let’s access the results of the centrality consensus analysis with the
function <code><a href="../reference/centrality_report.html">centrality_report()</a></code>:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/centrality_report.html">centrality_report</a></span><span class="op">(</span><span class="va">c_obj</span><span class="op">)</span></span>
<span><span class="co">#&gt;          candidate  degree betweenness closeness eigenvector central for</span></span>
<span><span class="co">#&gt; BC004004  BC004004 0.1667*     0.0719*       0.6          1*         dbe</span></span>
<span><span class="co">#&gt; Phe            Phe  0.1111      0.0065        1*           0           c</span></span></code></pre></div>
<div class="section level3">
<h3 id="copula-based-transfer-to-the-normal-realm">Copula-based transfer to the normal realm<a class="anchor" aria-label="anchor" href="#copula-based-transfer-to-the-normal-realm"></a>
</h3>
<p>Carmon harnesses the raw statistical behavior of the non-normally
distributed omics data to perform a proper transfer to the normal realm.
The statistical tool the package uses for this task are Gaussian
copulas. The first ingredient to perform this transition with copulas is
the choice of what statistical behavior best describes your omics data,
by deciding which marginal distribution is followed by each omics
layer.</p>
<p>The most basic approach to this is to choose the empirical
distribution, with which we take the data as they are, without making
any assumption. The approach carmon prefers, when possible, is to assume
that the data follow specific parametric statistical distributions. With
this approach, we use the omics data to estimate the parameters of such
distribution, allowing to find the distribution of the chosen family
that best describes the omics data.</p>
<p>With specific types of omics data, over the years, the scientific
community has settled on specific distributional assumptions. For
example, RNA-seq data, together with other omics types that generate
count-based measures, are generally assumed to follow a Negative
Binomial distribution. Carmon takes several of this known distributional
assumptions and implements them as a default to tailor its treatment of
those specific omics types. For all those omics types that have no known
assumption yet, or those for which we have not tailored a default
behavior yet, carmon chooses the first approach, the one based on the
assumption-free empirical distribution. To see for which omics types we
tailored a default behavior, please use the function
<code><a href="../reference/which_omics.html">which_omics()</a></code>. Be aware that carmon is built for being
upgraded, so soon there will be new omics types for which we implement a
tailored default behavior.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/which_omics.html">which_omics</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; "rna-seq", also as "rnaseq", "rna", "gene counts", "transcriptomics"</span></span>
<span><span class="co">#&gt;     is modeled by default as count data with a negative-binomial marginal.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; "mirna-seq", also as "mirnaseq", "microrna-seq", "micrornaseq", "mirna",</span></span>
<span><span class="co">#&gt;     "microrna"</span></span>
<span><span class="co">#&gt;     is modeled by default as count data with a negative-binomial marginal.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; "proteomics", also as "protein fragments", "protein counts"</span></span>
<span><span class="co">#&gt;     is modeled by default as count data with a negative-binomial marginal.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; "metabolomics", also as "lc-ms", "gc-ms", "ms"</span></span>
<span><span class="co">#&gt;     is modeled by default as positive continuous data with a log-normal</span></span>
<span><span class="co">#&gt;     marginal.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Anything else is modeled with the empirical marginal.</span></span></code></pre></div>
<p>These are also the terms you can use to name the elements of the
input R list if you want to trigger the default behavior for the
corresponding layers. For example, you could see that the two elements
in the input list <code>multi_omics_small</code> are
<code>"rnaseq"</code>, which will trigger the default assumption of
Negative Binomial distribution to model gene counts, and
<code>"metabolomics"</code>, which will trigger by default the
log-normal distribution to describe the metabolites’ concentrations.</p>
<p>If this is the default behavior for such omics types, but you know
that your particular data set follows a different statistical
distribution, you can run the function <code><a href="../reference/which_marginals.html">which_marginals()</a></code>.
This function displays which are the statistical distributions that
carmon is currently compatible with.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/which_marginals.html">which_marginals</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; "e" or "empirical" for using the empirical marginal distribution;</span></span>
<span><span class="co">#&gt; "n" or "normal" for using the normal marginal distribution;</span></span>
<span><span class="co">#&gt; "ln" or "lognormal" for using the log-normal marginal distribution;</span></span>
<span><span class="co">#&gt; "nb" or "negative binomial" for using the log-normal marginal distribution.</span></span></code></pre></div>
<p>This information can be used to customize the default behavior of
carmon with the copula-based transition to normal data. The two
arguments to the main wrapping function <code><a href="../reference/carmon.html">carmon()</a></code> that you
can use for this are:</p>
<ul>
<li>
<code>omics</code>: to be provided as a vector of characters, by
default it is unnecessary, as the input list will have the omics types
in the name given to each element. We recommend its use especially when
for some reason the elements are not or cannot be named. An example of
it could be when two layers in the list measure the same omics type, so
you may want to explicitate it as an input with this argument. This is
how you can use it in our example
<code>omics = c("rnaseq", "metabolomics")</code>.</li>
<li>
<code>marginals</code>: also a vector of characters, it overrides
the default behavior for the input omics types. It is also possible to
have a mixed case, in which for some omics layers you want to trigger
the default, and for others you want to customize it. If you know that
for your first layer you would rather use the assumption-free empirical
distribution, the normal for the second, and you prefer normal behavior
for the third, this is the form that the argument could take:
<code>marginals = c("e", "n", 0)</code>, requesting default behavior
with the <code>0</code>.</li>
</ul>
</div>
<div class="section level3">
<h3 id="network-reconstruction-and-selection">Network reconstruction and selection<a class="anchor" aria-label="anchor" href="#network-reconstruction-and-selection"></a>
</h3>
<p>Once the omics layers have been transferred to the normal realm,
carmon() proceeds to build the multi-omics network. For this task the
package implements multiple network reconstruction strategies, of which
one is chosen. By defualt, the package chooses <em>collaborative
graphical lasso</em> (<em>coglasso</em>) (<a href="#ref">Albanese,
Kohlen and Behrouzi, 2024</a>) as its engine to reconstruct the
multi-omics network, as the method was specifically developed for the
multi-omics network reconstruction scenario. carmon also implements
three classic alternatives to the reconstruction of networks:
<em>graphical lasso</em> (<a href="#ref">Friedman, Hastie and
Tibshirani, 2008</a>), <em>neighborhood selection</em> (<a href="#ref">Meinshausen and Buhlmann, 2006</a>), and <em>Pearson’s
correlation networks</em>. The network reconstruction strategy can be
personalized with the argument <code>net_method</code>, for example
<code>net_method = "coglasso"</code>, while for choosing one of the
other ones, respectively, you would need <code>"glasso"</code>,
<code>"mb"</code>, or <code>"correlation"</code>.</p>
<p>Apart from one specific use case, each of these methods explores
network reconstruction among a range of hyperparameters, usually
resulting in a different network for each different value (or
combination of values) these hyperparameters can take. This requires
specific model selection procedures that can pinpoint the best network
among the ones that have been built. Luckily, for each network
reconstruction method it implements, <code><a href="../reference/carmon.html">carmon()</a></code> also comes
with its tailored model selection procedures. For <em>coglasso</em>, for
example, it is possible to choose among three possible strategies. One
of them is the <em>extended Bayesian Information Criterion</em>
(<em>eBIC</em>). The other two are different versions of an algorithm
based on StARS (<a href="#ref">Liu, Roeder and Wasserman, 2010</a>),
extended to the <em>coglasso</em> case in which three different
hyperparameters need to be tuned, and described in (<a href="#ref">Albanese, Kohlen and Behrouzi, 2024</a>) as <em>XStARS</em>.
The two different versions implemented are indeed the original
<em>XStARS</em> and the more efficient but slightly less thorough
version <em>XEStARS</em>. By default, for <em>coglasso</em> as a network
reconstruction method <code><a href="../reference/carmon.html">carmon()</a></code> uses <em>XEStARS</em>. To
customize the model selection procedure, you can use the
<code>sel_method</code> parameter. For example, if instead of the
default you prefer the more thorough <em>XStARS</em>, you can set
<code>sel_method = "xstars"</code>, with the default being
<code>"xestars"</code> and <code>"ebic"</code> being the alternative for
using <em>eBIC</em>. For every other network reconstruction procedure,
as they are all based on a single hyperparameter, <code><a href="../reference/carmon.html">carmon()</a></code>
uses by default the one-dimensional <em>StARS</em>, but more options are
available. See them all in <code><a href="../reference/carmon.html">help(carmon)</a></code>, under the
description of the <code>sel_method</code> argument.</p>
<p>All of the methods mentioned above, either for network reconstruction
or network selection, can be further tweaked by giving specific
arguments. For example, if using <em>coglasso</em> for reconstruction,
and wanting to investigate the hyperparameter space more thoroughly than
the default, one could specify the <em>coglasso</em>-specific parameters
with a higher number, say <code>nlambda_w = 20</code> (normally only 8
values are explored), <code>nlambda_b = 20</code> (only 8 values
normally), and <code>nc = 10</code> (only 5 normally). To see more about
these and other arguments one can tweak about <em>coglasso</em> network
reconstruction and selection, please see <code><a href="https://drquestion.github.io/coglasso/reference/bs.html" class="external-link">?coglasso::bs</a></code>. For
the other network reconstruction methods, please see
<code><a href="https://rdrr.io/pkg/huge/man/huge.glasso.html" class="external-link">?huge::huge.glasso</a></code> for <em>graphical lasso</em>,
<code><a href="https://rdrr.io/pkg/huge/man/huge.mb.html" class="external-link">?huge::huge.mb</a></code> for <em>neighborhood selection</em>, and
<code><a href="https://rdrr.io/pkg/huge/man/huge.ct.html" class="external-link">?huge::huge.ct</a></code> for <em>Pearson’s correlation</em>, and see
<code><a href="https://rdrr.io/pkg/huge/man/huge.select.html" class="external-link">?huge::huge.select</a></code> for their associated model selection
procedures.</p>
<p>One specific run mode of <code><a href="../reference/carmon.html">carmon()</a></code> does not require model
selection, because the user is directly setting a value required
hyperparameter. It is possible when using
<code>net_method = "correlation"</code> and setting one between
<code>cor_cutoff</code> and <code>cor_quant</code>. The first determines
the threshold value of absolute Pearson’s correlation below which the a
connection should be excluded, while the second determines the quantile
of top connections (per absolute value) that need to be included in the
network. For example, setting <code>cor_quant = 0.05</code> will include
in the network only the top 5% connections based on the absolute
Pearson’s correlation value.</p>
</div>
<div class="section level3">
<h3 id="centrality-consensus-analysis">Centrality consensus analysis<a class="anchor" aria-label="anchor" href="#centrality-consensus-analysis"></a>
</h3>
<p>Once <code><a href="../reference/carmon.html">carmon()</a></code> selects a final network, it proceeds with
performing a centrality consensus analysis. In this step, four different
centrality measures are computed for the nodes of the network, with the
idea that the larger the consensus among the different measures about a
specific omics feature, the more likely this feature is to be important
to the studied biological phenomenon. The four measures employed are
<em>degree</em> centrality, <em>betweenness</em> centrality,
<em>closeness</em> centrality, and <em>eigenvector</em> centrality. By
default all the four of them are used for the consensus, but this can be
personalized with the argument <code>c_measures</code>, using the first
letter of each measure you want to employ. For example, to use only the
first three measures, you can set <code>c_measures = "dbc"</code>, while
the default would be <code>"dbce"</code>.</p>
<div class="section level4">
<h4 id="determining-the-size-of-the-consensus">Determining the size of the consensus<a class="anchor" aria-label="anchor" href="#determining-the-size-of-the-consensus"></a>
</h4>
<p>It is also possible to personalize how many omics features will be
included in the consensus with two arguments:
<code>max_candidates_c_measures</code>, and
<code>quantile_c_measures</code>. Both arguments determine the number of
top candidates selected by each measure separately, <em>before</em>
finding the consensus among the measures. The first does it with the
absolute value of candidates, the other with a determined quantile of
top candidates. The default values for these two arguments are
<code>max_candidates_c_measures = 20</code> and
<code>quantile_c_measures = 0.05</code>, and the smallest between the
two will be used. The same principle holds true when the user gives
values for both arguments, but when only one of the two is given that
one only determines the amount of candidates for the consensus.</p>
<p>Please note that the centrality analysis is the first optional step,
and it can be turned off by setting the argument to
<code><a href="../reference/carmon.html">carmon()</a></code> <code>analyse = FALSE</code>.</p>
</div>
</div>
<div class="section level3">
<h3 id="plot-of-network-and-of-centrality-analysis-results">Plot of network and of centrality analysis results<a class="anchor" aria-label="anchor" href="#plot-of-network-and-of-centrality-analysis-results"></a>
</h3>
<p>The final functional step of <code><a href="../reference/carmon.html">carmon()</a></code> is the plotting
module. <code><a href="../reference/carmon.html">carmon()</a></code> can generate two plots, depending on
whether the centrality analysis is carried on or not. The first one it
will show when the analysis is performed is the measures of centrality
for the top candidates, separately for each measure. This means that, by
default, four panels will be generated when all the four measures found
central nodes, but when a measure fails to find central nodes, then its
panel will not be displayed.</p>
<p>The second plot generated is the selected multi-omics network,
enriched by the results of the centrality analysis when this one is
performed. When an omics feature is found to be central according to a
measure, the first letter of the measure is added at the end of the
displayed node label, and the node itself is plotted larger. This also
means that the larger the consensus, the larger the node will be. At the
bottom of the plot, a legend shows the color coding for the different
omics types of the network and the different edge types for the
different edge strengths, depicting how strong the connection is. The
multi-omics network plot can be personalized with three different
arguments to <code><a href="../reference/carmon.html">carmon()</a></code>:</p>
<ul>
<li>
<code>plot_hot_nodes</code> can be set to <code>FALSE</code> to turn
off the enrichment of the network based on the centrality analysis, and
it is turned on by default.</li>
<li>
<code>plot_node_labels</code> can be set to <code>FALSE</code> to
hide the node labels from the plot, which is particularly useful for
larger networks. When <code>plot_hot_nodes</code> is <code>TRUE</code>
(default behavior) and <code>plot_node_labels</code> is
<code>FALSE</code>, only the central nodes’ labels are displayed in the
plot.</li>
<li>
<code>hide_isolated</code> hides the unconnected nodes from the plot
of the multi-omics network, <code>TRUE</code> by default.</li>
</ul>
<p>Please note that also the plotting module can be turned off by
setting the argument to <code><a href="../reference/carmon.html">carmon()</a></code>
<code>plot = FALSE</code>.</p>
</div>
<div class="section level3">
<h3 id="custom-run-of-carmon">Custom run of <code>carmon()</code><a class="anchor" aria-label="anchor" href="#custom-run-of-carmon"></a>
</h3>
<p>In this run, we customize the default behavior of
<code><a href="../reference/carmon.html">carmon()</a></code> with the information provided above, this time
reconstructing the network from the largest data set provided by the
package, <code>multi_omics</code>. We will will use alternative
marginals for the RNA-seq layer, keeping the default for the metabolomic
layer. We will use the omics features in the normal-realm to build a
network with Pearson’s correlation, selecting only the top 5%
connections. We will perform the consensus centrality analysis with all
four measures, but we will select the top 2% for each measure to select
the candidates for consensus. Moreover, we will hide the name of those
nodes that are not found to be central. Finally, we will inspect the
result of the consensus centrality analysis.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">multi_omics</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">multi_omics</span><span class="op">)</span> <span class="co"># 162 genes, 76 metabolites, for 30 observations</span></span>
<span><span class="co">#&gt; List of 2</span></span>
<span><span class="co">#&gt;  $ rnaseq      : num [1:30, 1:162] 407 54 148 269 76 91 500 482 65 201 ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:30] "BXD100" "BXD101" "BXD73b" "BXD44" ...</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:162] "Plin4" "Arc" "Egr2" "Tekt4" ...</span></span>
<span><span class="co">#&gt;  $ metabolomics: num [1:30, 1:76] 317 208 255 282 362 ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:30] "BXD100" "BXD101" "BXD73b" "BXD44" ...</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:76] "Ala" "Arg" "Asn" "Cit" ...</span></span>
<span></span>
<span><span class="va">c_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/carmon.html">carmon</a></span><span class="op">(</span><span class="va">multi_omics</span>,</span>
<span>    marginals <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"e"</span>, <span class="fl">0</span><span class="op">)</span>, <span class="co"># Sets the empirical marginal distribution</span></span>
<span>    <span class="co"># for the first layer, and the default</span></span>
<span>    <span class="co"># distribution for the second</span></span>
<span>    net_method <span class="op">=</span> <span class="st">"correlation"</span>, <span class="co"># Selecting Pearson's correlation</span></span>
<span>    cor_quant <span class="op">=</span> <span class="fl">0.05</span>, <span class="co"># Selecting the top 5% of connections</span></span>
<span>    quantile_c_measures <span class="op">=</span> <span class="fl">0.02</span>, <span class="co"># Top 2% of the candidates</span></span>
<span>    <span class="co"># to determine consensus of centrality</span></span>
<span>    plot_node_labels <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Checking sample-matching, formatting the data,</span></span>
<span><span class="co">#&gt;     and other formalities....</span></span>
<span><span class="co">#&gt; Done!</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ****************Beginning copulization****************</span></span>
<span><span class="co">#&gt; Copulizing layer 1 of 2 (rnaseq)</span></span>
<span><span class="co">#&gt; Copulizing layer 2 of 2 (metabolomics)</span></span>
<span><span class="co">#&gt; ****************Copulization complete*****************</span></span>
<span><span class="co">#&gt; ***********Beginning network reconstruction***********</span></span>
<span><span class="co">#&gt; Reconstructing network with Pearson's correlation....</span></span>
<span><span class="co">#&gt; Correlation cutoff is 0.64002200550331</span></span>
<span><span class="co">#&gt; Network reconstructed!</span></span>
<span><span class="co">#&gt; ***********Network reconstruction complete************</span></span>
<span><span class="co">#&gt; **************Beginning network analysis**************</span></span>
<span><span class="co">#&gt; Computing centrality measures....</span></span>
<span><span class="co">#&gt; Centrality measures computed!</span></span>
<span><span class="co">#&gt; Compiling the report....</span></span>
<span><span class="co">#&gt; Report compiled!</span></span>
<span><span class="co">#&gt; **************Network analysis complete***************</span></span></code></pre></div>
<p><img src="carmon_files/figure-html/custom%20run-1.png" width="576"><img src="carmon_files/figure-html/custom%20run-2.png" width="576"></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="../reference/centrality_report.html">centrality_report</a></span><span class="op">(</span><span class="va">c_obj</span><span class="op">)</span></span>
<span><span class="co">#&gt;               candidate  degree betweenness closeness eigenvector central for</span></span>
<span><span class="co">#&gt; Xbp1               Xbp1 0.2532*      0.0226    0.4072          1*          de</span></span>
<span><span class="co">#&gt; Dnajb11         Dnajb11 0.2278*      0.0076    0.3965     0.9875*          de</span></span>
<span><span class="co">#&gt; Gfod1             Gfod1 0.2194*      0.0062    0.3908     0.9597*          de</span></span>
<span><span class="co">#&gt; Hsph1             Hsph1 0.2068*      0.0045    0.3886     0.9415*          de</span></span>
<span><span class="co">#&gt; Pdia6             Pdia6  0.1983      0.0032    0.3853     0.9342*           e</span></span>
<span><span class="co">#&gt; P4ha1             P4ha1 0.2068*      0.0054    0.3853      0.9007           d</span></span>
<span><span class="co">#&gt; Hexim1           Hexim1  0.1899     0.0604*    0.4159      0.8308           b</span></span>
<span><span class="co">#&gt; Lrrc29           Lrrc29  0.0042           0        1*           0           c</span></span>
<span><span class="co">#&gt; Vwa3a             Vwa3a  0.0042           0        1*           0           c</span></span>
<span><span class="co">#&gt; Hfe2               Hfe2  0.0042           0        1*           0           c</span></span>
<span><span class="co">#&gt; Rbm11             Rbm11  0.0042           0        1*           0           c</span></span>
<span><span class="co">#&gt; PC ae C40:2 PC ae C40:2  0.0717      0.0107   0.4257*           0           c</span></span>
<span><span class="co">#&gt; Arl4a             Arl4a   0.038     0.0745*      0.34      0.0542           b</span></span>
<span><span class="co">#&gt; Pllp               Pllp  0.0211     0.0969*    0.2804      0.0025           b</span></span>
<span><span class="co">#&gt; Opalin           Opalin  0.0084     0.0897*    0.2361       1e-04           b</span></span>
<span><span class="co">#&gt; Phe                 Phe  0.0422     0.0867*    0.2033           0           b</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="ref">References<a class="anchor" aria-label="anchor" href="#ref"></a>
</h2>
<p>Albanese, A., Kohlen, W., &amp; Behrouzi, P. (2024). Collaborative
graphical lasso (arXiv:2403.18602). <em>arXiv</em> <a href="https://doi.org/10.48550/arXiv.2403.18602" class="external-link uri">https://doi.org/10.48550/arXiv.2403.18602</a></p>
<p>Friedman, J., Hastie, T., &amp; Tibshirani, R. (2008). Sparse inverse
covariance estimation with the graphical lasso. <em>Biostatistics</em>,
9(3), 432–441. <a href="https://doi.org/10.1093/biostatistics/kxm045" class="external-link uri">https://doi.org/10.1093/biostatistics/kxm045</a></p>
<p>Meinshausen, N., &amp; Buhlmann, P. (2006). High-dimensional graphs
and variable selection with the Lasso. <em>Ann. Statist.</em>, 34(3),
1436-1462. <a href="https://doi.org/10.1214/009053606000000281" class="external-link uri">https://doi.org/10.1214/009053606000000281</a></p>
<p>Liu, H., Roeder, K., &amp; Wasserman, L. (2010). Stability Approach
to Regularization Selection (StARS) for High Dimensional Graphical
Models (arXiv:1006.3316). <em>arXiv</em> <a href="https://doi.org/10.48550/arXiv.1006.3316" class="external-link uri">https://doi.org/10.48550/arXiv.1006.3316</a></p>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.5.2 (2025-10-31)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 24.04.3 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] carmon_0.99.0    BiocStyle_2.38.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] SummarizedExperiment_1.40.0 gtable_0.3.6               </span></span>
<span><span class="co">#&gt;  [3] xfun_0.54                   bslib_0.9.0                </span></span>
<span><span class="co">#&gt;  [5] ggplot2_4.0.0               Biobase_2.70.0             </span></span>
<span><span class="co">#&gt;  [7] lattice_0.22-7              vctrs_0.6.5                </span></span>
<span><span class="co">#&gt;  [9] tools_4.5.2                 generics_0.1.4             </span></span>
<span><span class="co">#&gt; [11] stats4_4.5.2                parallel_4.5.2             </span></span>
<span><span class="co">#&gt; [13] tibble_3.3.0                pkgconfig_2.0.3            </span></span>
<span><span class="co">#&gt; [15] BiocBaseUtils_1.12.0        Matrix_1.7-4               </span></span>
<span><span class="co">#&gt; [17] RColorBrewer_1.1-3          S7_0.2.0                   </span></span>
<span><span class="co">#&gt; [19] desc_1.4.3                  S4Vectors_0.48.0           </span></span>
<span><span class="co">#&gt; [21] lifecycle_1.0.4             compiler_4.5.2             </span></span>
<span><span class="co">#&gt; [23] farver_2.1.2                textshaping_1.0.4          </span></span>
<span><span class="co">#&gt; [25] DESeq2_1.50.0               Seqinfo_1.0.0              </span></span>
<span><span class="co">#&gt; [27] codetools_0.2-20            htmltools_0.5.8.1          </span></span>
<span><span class="co">#&gt; [29] sass_0.4.10                 yaml_2.3.10                </span></span>
<span><span class="co">#&gt; [31] pkgdown_2.1.3               pillar_1.11.1              </span></span>
<span><span class="co">#&gt; [33] jquerylib_0.1.4             BiocParallel_1.44.0        </span></span>
<span><span class="co">#&gt; [35] DelayedArray_0.36.0         cachem_1.1.0               </span></span>
<span><span class="co">#&gt; [37] abind_1.4-8                 locfit_1.5-9.12            </span></span>
<span><span class="co">#&gt; [39] tidyselect_1.2.1            digest_0.6.37              </span></span>
<span><span class="co">#&gt; [41] dplyr_1.1.4                 bookdown_0.45              </span></span>
<span><span class="co">#&gt; [43] fastmap_1.2.0               grid_4.5.2                 </span></span>
<span><span class="co">#&gt; [45] cli_3.6.5                   SparseArray_1.10.1         </span></span>
<span><span class="co">#&gt; [47] magrittr_2.0.4              S4Arrays_1.10.0            </span></span>
<span><span class="co">#&gt; [49] withr_3.0.2                 scales_1.4.0               </span></span>
<span><span class="co">#&gt; [51] rmarkdown_2.30              XVector_0.50.0             </span></span>
<span><span class="co">#&gt; [53] matrixStats_1.5.0           igraph_2.2.1               </span></span>
<span><span class="co">#&gt; [55] ragg_1.5.0                  evaluate_1.0.5             </span></span>
<span><span class="co">#&gt; [57] knitr_1.50                  GenomicRanges_1.62.0       </span></span>
<span><span class="co">#&gt; [59] IRanges_2.44.0              MultiAssayExperiment_1.36.0</span></span>
<span><span class="co">#&gt; [61] rlang_1.1.6                 Rcpp_1.1.0                 </span></span>
<span><span class="co">#&gt; [63] glue_1.8.0                  BiocManager_1.30.26        </span></span>
<span><span class="co">#&gt; [65] BiocGenerics_0.56.0         jsonlite_2.0.0             </span></span>
<span><span class="co">#&gt; [67] R6_2.6.1                    MatrixGenerics_1.22.0      </span></span>
<span><span class="co">#&gt; [69] systemfonts_1.3.1           fs_1.6.6                   </span></span>
<span><span class="co">#&gt; [71] coglasso_1.1.0</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Alessio Albanese, Pariya Behrouzi.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
